python/hidet/graph/ops/definitions/reduce/reduce_f16.py

```diff
@@ -0,0 +1,298 @@
+        if self.keep_dim:
+            remain_shape = [v if i not in dims else 1 for i, v in enumerate(shape)]
+        else:
+            remain_shape = [v for i, v in enumerate(shape) if i not in dims]
+        reduce_shape = [shape[i] for i in dims]
+        reduce_extent = prod(reduce_shape)
+        shape_32bit = [s // 2 if i == len(shape) - 1 else s for i, s in enumerate(shape)]
+        remain_layout = spatial(*remain_shape)
+
+        spatial_shape = []
+        repeat_shape = []
+        for i in range(len(shape_32bit)):
+            if i == len(shape_32bit) - 1:
+                spatial_shape.append(warp_size)
+                repeat_shape.append((shape_32bit[i] + warp_size - 1) // warp_size)  # num warps per row
+            elif i in dims:
+                spatial_shape.append(1)
+                repeat_shape.append(shape_32bit[i])
+            else:
+                spatial_shape.append(shape_32bit[i])
+                repeat_shape.append(1)
+        task_layout = repeat(*repeat_shape) * spatial(*spatial_shape)
+        grid_size = remain_layout.num_workers
+        accumulate_dtype = self.attributes['accumulate_dtype']
+        reduce_type = self.attributes['reduce_type']
+        ro = ReduceOperation.from_name(reduce_type)
+
+        with hidet.script_module() as module:
+
+            @hidet.script
+            def reduce_kernel(x: f16[x.const_shape()], y: f16[y.const_shape()]):
+                attr.cuda_grid_dim = grid_size
+                attr.cuda_block_dim = block_size
+                attr.cuda_min_blocks = 1
+
+                x_f32 = tensor_pointer('float32', shape=shape_32bit)
+                x_f32 = x
+
+                reg32 = register_tensor(f32, [1])
+                regs16 = tensor_pointer('float16', shape=[2])
+                regs16 = reg32
+                rv = register_tensor(accumulate_dtype, [1])
+                rv[0] = ro.initial_value(data_type(accumulate_dtype))
+                for indices in task_layout.on(threadIdx.x + blockIdx.x * block_size):
+                    reg32[0] = x_f32.read(indices, protected=True)
+                    rv[0] = ro.combine(rv[0], regs16[0])
+                    rv[0] = ro.combine(rv[0], regs16[1])
+                # Warp reduce by shuffle down
+                mask = active_mask()
+                rv[0] = ro.combine(rv[0], shfl_down_sync(mask, rv[0], 16, 32))
+                rv[0] = ro.combine(rv[0], shfl_down_sync(mask, rv[0], 8, 32))
+                rv[0] = ro.combine(rv[0], shfl_down_sync(mask, rv[0], 4, 32))
+                rv[0] = ro.combine(rv[0], shfl_down_sync(mask, rv[0], 2, 32))
+                rv[0] = ro.combine(rv[0], shfl_down_sync(mask, rv[0], 1, 32))
+                rv[0] = shfl_sync(mask, rv[0], 0, 32)
+                rv[0] = ro.finalize(acc=rv[0], size=reduce_extent)
+                if threadIdx.x == 0:
+                    for indices in remain_layout.on(blockIdx.x):
+                        y.write(indices, rv[0], protected=True)
```

Comment:
In this case, seems we do not need 'protected=True' as the number of thread blocks is exactly set to the required number of workers. Using protected write would add an condition statement to protect the write, one our bound analyzer can not remove it for us, there might be some performance degragation. 

@hjjq Could you double check whether we need this? Thanks!